src/samplers/gibbs.jl

```diff
@@ -71,51 +71,37 @@ function sample(model::Function, alg::Gibbs)
 
     for local_spl in spl.info[:samplers]
       if haskey(spl.info, :progress) local_spl.info[:progress] = spl.info[:progress] end
-      # dprintln(2, "Sampler stepping...")
+
       dprintln(2, "$(typeof(local_spl)) stepping...")
-      # println(varInfo)
-      if isa(local_spl.alg, Hamiltonian)
 
-        for _ = 1:local_spl.alg.n_iters
-          dprintln(2, "recording old θ...")
-          old_vi = deepcopy(varInfo)
-          is_accept, varInfo = step(model, local_spl, varInfo, i==1)
-          if ~is_accept
-            # NOTE: this might cause problem if new variables is added to VarInfo,
-            #    which will add new elements to vi.idcs etc.
-            varInfo = old_vi
-          end
-          if ~spl.alg.thin
-            samples[i_thin].value = Sample(varInfo).value
-            i_thin += 1
-          end
+      if isa(local_spl.alg, GibbsComponent)
+        if isa(local_spl.alg, PG)
+          # Clean possible wrong counter from HMC
+          # NOTE: can we move this into PG step? (or we actually don't need it)
+          varInfo.index = 0; varInfo.num_produce = 0
         end
-      elseif isa(local_spl.alg, PG)
-        # Update new VarInfo to the reference particle
-        varInfo.index = 0; varInfo.num_produce = 0
-        if ref_particle != nothing ref_particle.vi = varInfo end
 
-        # Clean variables belonging to the current sampler
-        varInfo = deepcopy(varInfo); varInfo[getretain(varInfo, 0, local_spl)] = NULL
-
-        # Local samples
         for _ = 1:local_spl.alg.n_iters
-          ref_particle, _ = step(model, local_spl, varInfo, ref_particle)
+          dprintln(2, "recording old θ...")
+          varInfo = step(model, local_spl, varInfo, i==1)
           if ~spl.alg.thin
-            samples[i_thin].value = Sample(ref_particle.vi).value
+            if isa(local_spl.alg, Hamiltonian) && ~local_spl.info[:accept_his][end]
```

Comment:
can we merge these lines 

```julia
if isa(local_spl.alg, Hamiltonian) && ~local_spl.info[:accept_his][end]
              samples[i_thin] = samples[i_thin - 1]
            else
              samples[i_thin].value = Sample(varInfo).value
            end
```

==>

```julia
     samples[i_thin].value = Sample(varInfo).value
```

Since if a proposal is rejected, `step` will return the old `VarInfo` anyway.